File: DataStructure/leetcode1124/solveLeetcode1124_method1.cpp
// class Solution {

// public:

//     int longestWPI(vector<int> &hours) {

//         int n = hours.size(), ans = 0, s = 0, pos[n + 2]; // 记录前缀和首次出现的位置

//         memset(pos, 0, sizeof(pos));

//         for (int i = 1; i <= n; ++i) {

//             hours[i - 1] > 8 ? --s : ++s; // 所有和前缀和相关的都取反

//             if (s < 0) ans = i;

//             else {

//                 if (pos[s + 1]) ans = max(ans, i - pos[s + 1]);

//                 if (pos[s] == 0) pos[s] = i;

//             }

//         }

//         return ans;

//     }

// };

File: DataStructure/leetcode1124/solveLeetcode1124_method1.go

// func longestWPI(hours []int) int {

// 	n := len(hours)

// 	ans := 0

// 	s := make([]int, n+1) // 前缀和

// 	st := []int{}

// 	st = append(st, 0)

// 	for j := 1; j <= n; j++ {

// 		if hours[j-1]>8{
//             s[j]=s[j-1]+1
//         }else{
//             s[j]=s[j-1]-1
//         }

// 		if s[j] < s[st[len(st)-1]] {

// 			st = append(st, j) // 感兴趣的 j

// 		}

// 	}

// 	for i := n; i > 0; i-- {

// 		for len(st) > 0 && s[i] > s[st[len(st)-1]] {

// 			ans = int(max(ans, i-st[len(st)-1])) // [栈顶,i) 可能是最长子数组

// 			st = st[:len(st)-1]

// 		}

// 	}

// 	return ans

// }